riscv-c-api.md

```diff
@@ -116,5 +116,41 @@ For example:
 
 ## Intrinsic Functions
 
-Do we really have none of these?  I can't figure out
-`gcc/gcc/config/riscv/riscv-builtins.c`...
+RISC-V intrinsics allow access to RISC-V-specific instructions from C or C++ that are usually not otherwise synthesizable by compilers.
+Intrinsics typically map to a single instruction that is inlined to avoid function call overhead.
+
+Intrinsics are only available if the corresponding instructions are available (as part of the effective `-march` string).
+Some intrinsics are only available if a particular header file is included.
+RISC-V header files, that enable intrinsics, require the prefix `riscv_` (e.g. `riscv_vector.h` or `riscv_crypto.h`).
+
+RISC-V intrinsics follow the following naming rule:
+
+```
+INTRINSIC ::= PREFIX '_' MNEMONIC [ '_' RET_TYPE ]
+PREFIX ::= "__builtin_riscv" | "__rv" EXTENSION | "_rv" | "_rv32" | "_rv64"
+EXTENSION ::= ISA extension letter or string.
+MNEMONIC ::= Instruction name in the ISA extensions specification. Replace '.' with '_' (if any).
+RET_TYPE ::= Return type indication.
+```
+
+The return type is typically encoded into the intrinsic name as part of the prefix or a postfix.
+In case the return type is encoded into the prefix, then the following prefixes should be used:
+* `_rv`: for intrinsics that operate on the long data type.
+* `_rv32`: for intrinsics that operate on the int32_t data type.
+* `_rv64`: for RV64-only intrinsics that operate on the int64_t data type.
+
+The `PREFIX` has the intention to avoid naming collisions.
+RISC-V header files can provide aliases to the intrinsics without the prefix.
+
+```
+RISC-V intrinsics examples:
+
+orc.b rd, rs:
+int32_t __builtin_riscv_orc_b_32(int32_t rs);
+
+vadd.vv vd, vs2, vs1:
+vint8m1_t __rvv_vadd_vv_i8m1(vint8m1_t vs2, vint8m1_t vs1, size_t vl);
```

Comment:
Actually we define vector intrinsic without prefix now (e.g. `vadd_vv_i8m1`), and it's already implemented on upstream LLVM for a while and used in several open source projects like OpenCV.

Reply:
I agree, my team has written and shipped thousands of lines using the current API.

But a few lines up it says, 
>The `PREFIX` has the intention to avoid naming collisions.
RISC-V header files can provide aliases to the intrinsics without the prefix.

So maybe this won't affect me? (And can this aliasing be done without further bloat?)

Reply:
Yeah, we cannot pull back published APIs.
The intent is that we standardize the naming so that we have a common pattern across all the intrinsics for ISA extensions.

So yes, the example is wrong. Should be `vint8m1_t vadd_vv_i8m1(...)` after including `riscv_vector.h`, right?

What about the type prefix `__rvv_` (e.g. `__rvv_int8m1_t`) as defined in https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/RISCVVTypes.def and the function prefix `__builtin_rvv_` as defined in https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/riscv_vector.td#L114, i.e. would this be a reasonable prefix pattern?

Reply:
The riscv_vector.h file in LLVM is currently somewhere around 8-10MB containing 10s of thousands of intrinsics and is quite slow to parse. I worry that adding aliases will make this even larger and increase the already slow compile time.

Reply:
Maybe we could define the common rule for intrinsic header first (https://github.com/riscv-non-isa/riscv-c-api-doc/issues/17), and those new naming rule API are only supported in new common intrinsic headers.
For now, toolchain still need to support old intrinsic header with old ABI, and they will be deprecated in the future.
In this way, I think we could get rid of the current API designed and make things moving forward.

Reply:
> Maybe we could define the common rule for intrinsic header first (#17), and those new naming rule API are only supported in new common intrinsic headers. For now, toolchain still need to support old intrinsic header with old ABI, and they will be deprecated in the future. In this way, I think we could get rid of the current API designed and make things moving forward.

On file naming, how about this ( already in the prototype/algorithm test repo https://github.com/rvkrypto/rvkrypto-fips )
* To unify things (with vector), the main name for the crypto extension is  `riscv_crypto.h`. With version 1.0 this just pulls in `risc_crypto_scalar.h` as the first ratified version didn't specify vector crypto exentsoins.
* Once vector cryptography is defined, this `riscv_crypto.h` can pull in both `riscv_crypto_scalar.h` and `riscv_crypto_vector.h`. Only the latter latter has a dependency on the vector extension, and there are reasons why one might not want to include it in all source modules that just need a few bits from the scalar crypto.
* Once bitmanip intrinsics are defined again, `riscv_crypto_scalar.h` can pull in the relevant `riscv_bitmanip_.._.h` files, as the scalar cryptography extension actually includes a set of bitmanip instructions, which we currently define "for them". One can't define those twice.
* I guess a unified API rule is what this PR is about, I basically just need the rule to be fixed for this to go forward, So currently I'm just waiting for @cmuellner to signal that names are fixed, and then I'll pull the trigger and rename all of the intrinsics accordingly.
* At least for scalar instructions, It seems best to consider builtins to be a compiler-specific implementation mechanism for intrinsics. Intrinsics are enabled by including the relevant header file. The underscore prefix is reserved for the compiler/system namespace, and I don't want the scalar header to cause namespace collisions. However for vector intrinsics, for convenience purposes, the prefix may be dropped I guess. 

Some historical comments:
* Name of the file aside, it's a shame that the Bitmanip file `rvintrin.h` got blocked due to FSF copyright assignment paperwork (I have the FSF paperwork, if that matters).  Its prefix naming convention `_rv`/`_rv32`/`_rv64` based on the variable/return type is adequate for scalar ISA extensions. Various quirks related to RISC-V sign extension conventions and some other things are handled by it nicely; starting from scratch would require those things to be re-learned. It's easier just to change names.
* Anyway, rather than including the bitmanip header file, I had to rewrite the selected parts of the bitmanip header. This wasn't too hard since I didn't rewrite the emulation parts, just refactored them out. Emulation should not be done by a header like this (specifically in crypto as it destroys the constant-time properties asserted.)
* The P extension is not frozen, nor is it in the imminent ratification roadmap. I guess this comes down to it in some ways "competing" with the vector extension. So it can have relatively low priority from RISC-V perspective. However, there is a vendor or two that support it, so perhaps the compiler people can just ignore RISC-V and consider the P extension coming from the vendor instead. But it shouldn't be in *this* spec before it moves forward in RVI.
* Scalar crypto has been ratified, as have parts of bitmanip. These are in silicon tape-outs from multiple vendors. Since intrinsics are the main C programming method specifically for scalar crypto, we'd like to get this fixed soon.

Reply:
> So yes, the example is wrong. Should be vint8m1_t vadd_vv_i8m1(...) after including riscv_vector.h, right?
>
> What about the type prefix __rvv_ (e.g. __rvv_int8m1_t) as defined in https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/RISCVVTypes.def and the function prefix __builtin_rvv_ as defined in https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/riscv_vector.td#L114, i.e. would this be a reasonable prefix pattern?

I suppose user should never use those internal stuffs like `__rvv_int8m1_t`, that is not portable across different compiler implementation.

And user should not expect those intrinsic functions are available without including header, always import that might increase lots of compilation time, especially something like vector intrinsic functions.

Reply:
I prefer `vadd_vv_i8m1` to `__rvv_vadd_vv_i8m1`. The RVV GCC implementation doesn't  support `__rvv_vadd_vv_i8m1` bu support `vadd_vv_i8m1`. Because it implement RVV intrinsics using `simulate_decl` like `ARM SVE`. The RVV intrinsisc header 
`riscv_vector.h` in https://github.com/riscv-collab/riscv-gcc/blob/riscv-gcc-rvv-next/gcc/config/riscv/riscv_vector.h does not add any aliases and it can reduce compilation time. I personally doesn't like the api that has prefix `_rvv` and need alias added in the header for RVV intrinsics.